pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -106,27 +105,47 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 	err = r.client.Get(context.TODO(), types.NamespacedName{Name: buildRun.Spec.BuildRef.Name, Namespace: buildRun.Namespace}, build)
 	if err != nil {
 		reqLogger.Error(err, "Failed to get Build from BuildRun", "Build", buildRun.Spec.BuildRef.Name)
+		buildRun = r.updateBuildRunErrorStatus(buildRun, err)
```

Comment:
I notice the action of updating the resource status is frequently duplicated, should we extract that in a single routine?

Reply:
We catch each failure and update the status by using each error. so the errors are different.

The status update won't be executed multiple times, the first failure will cause to update status then reconcile.

I am not sure if I answer your question?

Reply:
What I meant is those lines [here](https://github.com/zhangtbj/build/blob/264a9043bb9e56bddcb26f6ff8540a6c9dc8530e/pkg/controller/buildrun/buildrun_controller.go#L107-L113), [here](https://github.com/zhangtbj/build/blob/264a9043bb9e56bddcb26f6ff8540a6c9dc8530e/pkg/controller/buildrun/buildrun_controller.go#L118-L124), [here](https://github.com/zhangtbj/build/blob/264a9043bb9e56bddcb26f6ff8540a6c9dc8530e/pkg/controller/buildrun/buildrun_controller.go#L143-L149), and in other places are almost the same routine. Therefore, should we extract that routine as a `func`?

Reply:
Im currently reviewing the PR, for this particular discussion, there is actually an issue on the following:
```go
	if err != nil {
		reqLogger.Error(err, "Failed to get Build from BuildRun", "Build", buildRun.Spec.BuildRef.Name)
		buildRun = r.updateBuildRunErrorStatus(buildRun, err)
		updateErr := r.client.Status().Update(context.TODO(), buildRun)
		if updateErr != nil {
			return reconcile.Result{}, updateErr
		}
		return reconcile.Result{}, err
	}
```
the above is managing 2 errors, but only one error is returned. I think what you can do here is:
- return two errors in `return reconcile.Result{}, err` , `updateErr` might be `nil` or `populated` but as long as you can return both, is much better. Even doing something like `fmt.Errorf("errors: %v %v", err, updateErr)` can do the job for now.

For what @otaviof is saying I think is a good point, you can do this by putting the client calls inside `updateBuildRunErrorStatus()` to avoid those repetitive code lines.

Reply:
Hi @otaviof and @qu1queee ,

Yes I got your point and refine the logic so that most of the common update logic are extracted in the `updateBuildRunErrorStatus` func and yes, I also noticed the duplicated status update.

I have refined and fixed in the second commit.

Reply:
I just dropped you more comments, sorry I was in the middle of checking the PR, take your time

Reply:
Hi @otaviof and @qu1queee ,

Great thanks for your comments and suggestions. I have refined them and should address all of the issues.

I also verified the e2e tests by using the latest commits, it passed:
```
✔ $ operator-sdk test local ./test/e2e --go-test-flags "-v -timeout=20m" --up-local --namespace default
INFO[0000] Testing operator locally.                    
time="2020-04-26T17:30:47+08:00" level=info msg="Started local operator"
=== RUN   TestBuild
=== RUN   TestBuild/build-group
=== RUN   TestBuild/build-group/Build_e2e_tests
=== PAUSE TestBuild/build-group/Build_e2e_tests
=== CONT  TestBuild/build-group/Build_e2e_tests
--- PASS: TestBuild (769.89s)
    --- PASS: TestBuild/build-group (0.00s)
        --- PASS: TestBuild/build-group/Build_e2e_tests (641.45s)
...
ok      github.com/redhat-developer/build/test/e2e      803.591s
INFO[0806] Local operator test successfully completed. 
```

Pls let me know if you have other suggestion. Thanks!